lib/Sema/ConstraintLocator.cpp

```diff
@@ -473,6 +475,10 @@ void ConstraintLocator::dump(SourceManager *sm, raw_ostream &out) const {
     case Condition:
       out << "condition expression";
       break;
+
+    case ForEachAuxiliary:
```

Comment:
I'd be very happy if we didn't do it this way (which effectively hides the problem) and instead addressed a root case limitation of the constraint system. I feel like this is yet another case when diagnostics are put on the back-burner so somebody else would have deal with that later. I'd like to propose an alternative approach.

The problem here is related to the fact that there is no way to express dependence of `makeIterator` on its base type conforming to `Sequence`. There is, maybe not the most elegant way, to fix that limitation in the current framework by changing how `conforms to` constraint behaves or introducing a new similar constraint which would eliminate a need for special locator because it would be possible to record a non-confirming type as a constraint system "hole". 

`Conforms To` constraint could be extended to support a third type which gets bound when conformance is "confirmed" as a result of successful constraint simplification, this could also be expressed as a separate `Conforms To (Bind)` constraint, such would be used to establish ordering between conformance constraint and its dependents e.g. `T0 <conforms to and bind> Sequence -> T1`, resulting `T1` could be used as a base type for member lookup e.g. `T1[.makeIterator] == T2`. `T1` would only be bound if `T0` conforms to `Sequence` which means that `.makeIterator` constraint cannot be simplified until conformance is checked (because its base type wouldn't be resolved). This also gives the solver an opportunity to mark `T1` as a "hole" if `T0` doesn't conform and just drop `.makeIterator` constraint when needed without having to produce any fixes.

Reply:
I've reimplemented the last commit with the new "value witness" constraint that does essentially what you describe.